### ISSUE TITLE
Predicting significant CCIs

### ISSUE THREAD

**Issue by sailseem** [BiomedicalMachineLearning/stLearn#160]
Dear developer,

Thank you for developing such a great package. Well, I have one small question when I run the 
st.tl.cci.run_cci(data, 'cell_type', 
                  min_spots=3, 
                  spot_mixtures=True, 
                  cell_prop_cutoff=0.2, 
                  sig_spots=True, 
                  n_perms=100 
                 )

Exception: Cell type scores misformatted in adata.uns[cell_type]:
Cell types missing from adata.uns[cell_type] columns:
['AC-like Malignant', 'ECs', 'MES-like Malignant', 'NKT', 'malignant', 'Olo', 'EC', 'macrophage']

Can you help me with it?
Thanks

**Comment by BradBalderson** [comment]
Hi @sailseem,

Thanks alot for running st.tl.cci.run_cci!

Sure can help you. That error means the expected inputted state of the adata object was not met.

[In the tutorial](https://stlearn.readthedocs.io/en/latest/tutorials/stLearn-CCI.html), it is this part:
```
# Adding the label transfer results,  #
spot_mixtures = pd.read_csv(data_dir+'../../Brad/label_transfer_bc.csv', index_col=0, sep='\t')
labels = spot_mixtures.loc[:,'predicted.id'].values.astype(str)
spot_mixtures = spot_mixtures.drop(['predicted.id','prediction.score.max'],
                                   axis=1)
spot_mixtures.columns = [col.replace('prediction.score.', '')
                         for col in spot_mixtures.columns]

# Note the format! #
print(labels)
print(spot_mixtures)
print('Spot mixture order correct?: ',
      np.all(spot_mixtures.index.values==data.obs_names.values)) # Check is in correct order

# NOTE: using the same key in data.obs & data.uns
data.obs['cell_type'] = labels # Adding the dominant cell type labels per spot
data.obs['cell_type'] = data.obs['cell_type'].astype('category')
data.uns['cell_type'] = spot_mixtures # Adding the cell type scores
```

It appears from the error that, the data frame stored in adata.uns['cell_type'], is missing columns for cell type which are present in adata.obs['cell_type']. 

In other words, my best guess is:
`np.all([cell_type in adata.uns['cell_type'].columns for cell_type in adata.obs['cell_type`].cat.categories])`
is False.

**Comment by sailseem** [comment]
Well, i have to use sep=',' to read my csv file, using sep='\t' cannot apart from each column.
And this code returns like this, which 0 column, I don't know why
![image](https://user-images.githubusercontent.com/72428172/162340679-a6e7e792-d8d1-4e92-baa2-b87419cafc62.png)

here is other returns,
![image](https://user-images.githubusercontent.com/72428172/162340738-4fb658ba-bc7e-4418-acf9-3a0e9ce59b2f.png)
And my prediction csv file looks like this,
![image](https://user-images.githubusercontent.com/72428172/162340780-3f7d0716-955a-4c99-b4cb-4bee86365779.png)

**Comment by BradBalderson** [comment]
Got it. Few suggestions which might help:

- The predictions at the moment only include the predicted.id and the maximum label transfer score. If you want to use spot_mixture=True (i.e. model each spatial location as a mixture of cell types), need to have the scores of each cell type for each spot. So it should look like this:

```
                   basal_like_2  macrophage    stroma  basal_like_1  \
AAACAAGTATCTCCCA-1      0.114847    0.153870  0.517986      0.000000
AAACACCAATAACTGC-1      0.157768    0.004139  0.071592      0.451562
AAACAGAGCGACTCCT-1      0.202656    0.004635  0.647263      0.003862
AAACAGGGTCTATATT-1      0.000000    0.082156  0.003555      0.000000
AAACAGTGTTCCTGGG-1      0.565240    0.039651  0.205373      0.119639
```
Each column is a cell type, row is each spot/spatial location/cell, each value is the label transfer/deconvolution score for that cell type in that spot. 

- If you don't want to model each spatial location as a mixture of cell types, you can simply set spot_mixtures=False, and this information will not be required. So like this:

```
 st.tl.cci.run_cci(data, 'cell_type',
min_spots=3,
spot_mixtures=False, #This is false, so no longer need spot_mixture information in adata.uns!!!!!
cell_prop_cutoff=0.2, #This irrelevant if we aren't modelling spots as mixtures. 
sig_spots=True,
n_perms=100
)
```

- When looking at the text file, instead of opening in excel, perhaps open in a text editor (like textedit in mac, or notepad in windows). That way can see what the separator is, i'm guessing it's not a ',' since it's missing the columns when loaded in pandas.

**Comment by sailseem** [comment]
Thanks for your guidance. Well, when i set spot_mixtures=False, the kernel was dead. And when I run the st.pp.neighbors(data,n_neighbors=15,use_rep='X_pca',random_state=0), kernel dead as well.

Any ideas of that?

**Comment by BradBalderson** [comment]
Kernal being dead sounds like non-stlearn related issue. If you're running in jupyter notebook, was the notebook server shutdown at the terminal, but still trying to run?

**Comment by sailseem** [comment]
All right, thanks, no ,some bugs or compatibility issues. I will try to figure that out. Thanks

**Comment by BradBalderson** [comment]
Okay, if it does happen to be stlearn issue please let me know!
